File: system/app/core/isolation.py
"""Project-local paths for the Vanilla distribution."""
from pathlib import Path

from .layout import ROOT

def inside(root, value):
    root = Path(root).resolve()
    value = Path(value)
    resolved = (value if value.is_absolute() else root / value).resolve()
    if resolved == root or not resolved.is_relative_to(root):
        raise ValueError("Vanilla-Pfade müssen innerhalb des eigenen Projektordners liegen.")
    return resolved

def port(value):
    value = int(value)
    if not 1024 <= value <= 65535 or value in {8890, 9090}:
        raise ValueError("Unzulässiger Vanilla-Port.")
    return value

File: wrapper/isolation.mjs
import path from 'node:path';
import {existsSync, realpathSync} from 'node:fs';
import {fileURLToPath} from 'node:url';
export const projectRoot = realpathSync(fileURLToPath(new URL('../', import.meta.url)));
export function localPath(value, root = projectRoot) {
  root = realpathSync(root);
  const target = path.resolve(root, value);
  let ancestor = target;
  while (!existsSync(ancestor)) ancestor = path.dirname(ancestor);
  const resolved = path.resolve(realpathSync(ancestor), path.relative(ancestor, target));
  if (resolved === root || !resolved.startsWith(root + path.sep))
    throw new Error('Vanilla-Pfade müssen innerhalb des eigenen Projektordners liegen.');
  return resolved;
}
export function localPort(value) {
  const port = Number(value);
  if (!Number.isInteger(port) || port < 1024 || port > 65535 || [8890,9090].includes(port))
    throw new Error('Unzulässiger Vanilla-Port.');
  return port;
}
